src/estimation/src/AttitudeMahonyFilter.cpp

```diff
@@ -0,0 +1,231 @@
+{
+    iDynTree::Matrix3x3 out;
+    iDynTree::toEigen(out) = iDynTree::toEigen(a)*iDynTree::toEigen(b).transpose();
+    return out;
+}
+
+iDynTree::Matrix3x3 getAngVelSkewSymmetricMatrixFromMeasurements(iDynTree::Vector3 meas, const iDynTree::Direction& vectorDir, double confidenceMeas, const iDynTree::Rotation& R)
+{
+    iDynTree::Direction va_hat = R.inverse().changeCoordFrameOf(vectorDir);
+    iDynTree::Vector3 vectorial_estimate = iDynTree::Vector3(va_hat.data(), 3);
+    iDynTree::toEigen(meas).normalize();
+    iDynTree::Matrix3x3 A_acc = getMatrixFromVectorVectorMultiplication(meas, vectorial_estimate);
+    iDynTree::Matrix3x3 S_acc;
+    iDynTree::toEigen(S_acc) = (iDynTree::toEigen(A_acc) - iDynTree::toEigen(A_acc).transpose())*(confidenceMeas*0.5);
+    return S_acc;
+}
+
+bool iDynTree::AttitudeMahonyFilter::updateFilterWithMeasurements(const iDynTree::LinearAccelerometerMeasurements& linAccMeas, const iDynTree::GyroscopeMeasurements& gyroMeas)
+{
+    iDynTree::Matrix3x3 S_acc = getAngVelSkewSymmetricMatrixFromMeasurements(linAccMeas, m_gravity_direction, 1-m_params.confidence_magnetometer_measurements, m_orientationInSO3);
+    iDynTree::toEigen(m_omega_mes) = -iDynTree::toEigen(mapso3ToR3(S_acc));
+    m_Omega_y = gyroMeas;
+    return true;
+}
+
+bool iDynTree::AttitudeMahonyFilter::updateFilterWithMeasurements(const iDynTree::LinearAccelerometerMeasurements& linAccMeas, const iDynTree::GyroscopeMeasurements& gyroMeas, const iDynTree::MagnetometerMeasurements& magMeas)
+{
+    if (m_params.use_magnetometer_measurements == false)
+    {
+        iDynTree::reportWarning("AttitudeMahonyFilter", "updateFilterWithMeasurements", "useMagnetoMeterMeasurements set to false, using only accelerometer measurements");
+        return updateFilterWithMeasurements(linAccMeas, gyroMeas);
+    }
+
+    iDynTree::Matrix3x3 S_acc = getAngVelSkewSymmetricMatrixFromMeasurements(linAccMeas, m_gravity_direction, 1-m_params.confidence_magnetometer_measurements, m_orientationInSO3);
+    iDynTree::Matrix3x3 S_mag = getAngVelSkewSymmetricMatrixFromMeasurements(magMeas, m_earth_magnetic_field_direction, m_params.confidence_magnetometer_measurements, m_orientationInSO3);
+    iDynTree::Matrix3x3 S_meas;
+    iDynTree::toEigen(S_meas) = iDynTree::toEigen(S_acc) + iDynTree::toEigen(S_mag);
+
+    iDynTree::toEigen(m_omega_mes) = -iDynTree::toEigen(mapso3ToR3(S_meas));
+    m_Omega_y = gyroMeas;
+    return true;
+}
+
+bool iDynTree::AttitudeMahonyFilter::propagateStates()
+{
+    iDynTree::Vector3 gyroUpdate;
+    iDynTree::toEigen(gyroUpdate) = iDynTree::toEigen(m_Omega_y) - iDynTree::toEigen(m_state.m_gyroscope_bias) + (iDynTree::toEigen(m_omega_mes)*m_params.kp);
+
+    iDynTree::Quaternion correction = pureQuaternion(gyroUpdate);
+    iDynTree::toEigen(m_state.m_orientation) = iDynTree::toEigen(m_state.m_orientation) + iDynTree::toEigen((composeQuaternion2(m_state.m_orientation, correction)))*(m_params.time_step_in_seconds*0.5);
```

Comment:
If you want to simplify the code, you can omit the `iDynTree::` part, by using `using iDynTree::toEigen` at the start of the file or the function. `using  namespace iDynTree` is a bad practice due to the possible collision, but targeting `using`-directive may have sense if they simplify the code. 

Furthermore, if you need to map to Eigen an iDynTree vector or matrix multiple times in an expression, you can also define Eigen::Map local variables, such as: 
https://github.com/robotology/idyntree/blob/c8bf721b771fa4b1e7c3a940632e121060719a19/src/core/src/SpatialInertia.cpp#L84
or https://github.com/robotology/idyntree/blob/c8bf721b771fa4b1e7c3a940632e121060719a19/src/core/src/SpatialInertia.cpp#L153